server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1280,10 +1299,16 @@ protected InputStream openSlice(long slice) throws IOException {
                         }
                     }
                 }
-            }.restore(snapshotFiles, store);
-        } catch (Exception e) {
-            throw new IndexShardRestoreFailedException(shardId, "failed to restore snapshot [" + snapshotId + "]", e);
-        }
+            }.restore(snapshotFiles, store, l);
```

Comment:
I understand the changes around `FileRestoreContext` but this is adding more complexity on something that is already torn to fit for CCR and regular restore needs. I think we should look at simplify `FileRestoreContext` before adding async behavior to it. Or maybe you already investigated this?

Reply:
I found some significant simplifications here in https://github.com/elastic/elasticsearch/pull/48173 when investigating this for this PR but that was pretty much all the simplification I could find (it does remove incorrectly shared code between blobstore and CCR repo). Actually, with that change the separation between CCR and blobstore code should be "correct" in theory as all the code in the base calss is now used across both implementations.

I think I'd "simply" make the CCR logic properly async in a follow-up as that would simplify it overall as well but figured for now keeping CCR code as is and blocking around the changed API was the way to go to not mix things into this PR.